educator/src/Dscp/DB/SQLite/Functions.hs

```diff
@@ -182,170 +183,135 @@ closeSQLiteDB sd =
-        UnliftIO unlift <- DBT askUnliftIO
-        return $ UnliftIO $ unlift . runDBT
-
-instance MonadReader r m => MonadReader r (DBT t w m) where
-    ask = DBT $ lift ask
-    reader = DBT . lift . reader
-    local doModify = DBT . mapReaderT (local doModify) . runDBT
-
-instance (Log.MonadLogging m, Monad m) => Log.MonadLogging (DBT t w m) where
-    log = DBT . lift ... Log.log
-    logName = DBT $ lift Log.logName
-
-instance (Log.ModifyLogName m, Monad m) => Log.ModifyLogName (DBT t w m) where
-    modifyLogNameSel how = DBT . mapReaderT (Log.modifyLogNameSel how) . runDBT
-
-dbtToSqliteM :: MonadUnliftIO m => DBT t w m (DBT t w m a -> SqliteM a)
-dbtToSqliteM = do
-    UnliftIO unliftIO <- DBT $ lift askUnliftIO
-    return $ \(DBT action) ->
-        SqliteM . ReaderT $ \(_logger, conn) -> unliftIO (runReaderT action conn)
-
-sqliteMToDbt :: MonadIO m => SqliteM a -> DBT t w m a
-sqliteMToDbt (SqliteM action) =
-    DBT . ReaderT $ \conn -> liftIO $ runReaderT action (logger, conn)
-  where
-    logger _ = pass
-    -- logger = sqlDebugLogger
-
-instance MonadUnliftIO m =>
-         MonadBeam SqliteCommandSyntax Sqlite Connection (DBT t w m) where
-    -- We omit implementation of the following methods as soon as they work in IO only,
-    -- use dedicated 'DBT' runners instead ('runInsert' and others).
-    withDatabaseDebug = error "DBT.withDatabaseDebug: not implemented"
-    withDatabase = error "DBT.withDatabase: not implemented"
-    runNoReturn = error "DBT.runNoReturn: not implemented"
-    runReturningMany = error "DBT.runReturningMany: not implemented"
-
--- | Declares whether given 'DBT' actions should be performed within
--- transaction.
-data TransactionalContext = WithinTx | OutsideOfTransaction
-
--- | Declares whether given 'DBT' actions, if performed within transaction,
--- should be performed within /write/ transaction.
--- We cannot rely on sqlite itself detecting writing transactions, otherwise
--- "ErrorBusy: database is locked" errors are possible,
--- see https://stackoverflow.com/questions/30438595/sqlite3-ignores-sqlite3-busy-timeout
--- for details.
-data OperationType = Writing | Reading
+data TransactionContextProvided = TransactionContextProvided
+type WithinTx = Given TransactionContextProvided
+
+data WriteContextProvided = WriteContextProvided
+type WithinWrite = Given WriteContextProvided
+
+type WithinWriteTx = (WithinTx, WithinWrite)
+
+allowTxUnsafe :: (WithinTx => a) -> a
+allowTxUnsafe = give TransactionContextProvided
+
+allowWriteUnsafe :: (WithinWrite => a) -> a
+allowWriteUnsafe a = give WriteContextProvided a
+
+allowWriteTxUnsafe :: (WithinWriteTx => a) -> a
+allowWriteTxUnsafe a = allowWriteUnsafe $ allowTxUnsafe a
 
 ------------------------------------------------------------
 -- SQLite queries building
 ------------------------------------------------------------
 
-{- We rewrite runners as soon as it allows requiring write or transaction context.
+type MonadQuery m = (m ~ SqliteM)
```

Comment:
Aren't we into transferring to postgres?

Reply:
Yep, in the next PR.